### PR TITLE
Usability proposals

### PR DESCRIPTION
**Pull request by jamartinh**
Hi @takuseno I am proposing a new PR with some minor improvements for you to consider:

+ Training Loop now with tqdm auto and disable feature
+ New parameters shuffle (bool)  (default True) for allowing the user to decide whether to shuffle the transitions or not
+ Progress bar now displays, current epoch, and loss values during epoch and mini batch steps
+ Training Loop simplified by iterating on chunks (mini batches)

### CODE REVIEW

**Comment by takuseno** [comment]
@jamartinh Hello, thank you for your continuous contribution! For now, I'm not convinced to use `more-itertools` since it's only used in one place. But, I like several things you proposed here. So please focus on the following things.
- `shuffle` flag
- more descriptions at `tqdm` progress
  - I'd like you to paste screenshots to see how it looks like

Thank you!

**Comment by jamartinh** [comment]
On Jupyter notebooks looks like this:
image

On terminal looks like this:
image

**Comment by jamartinh** [comment]
I have tested shuffle flag and it is working as expected.
W.r.t. more-itertools there is the option of simply use another chunked implementation, it is very simple and there are many shared codes doing the same. I can do it easily.
However, more-itertools seems to be so standard that probably those recipes will be incorporated to the standard python module itertools in next releases.

**Comment by jamartinh** [comment]
We can stick to use numy implementation numpy.array_split(ary, indices_or_sections, axis=0) as well

**Comment by jamartinh** [comment]
1. removed more-itertools dependency
2. reworked chunk generation with pure python code
3. reworded shuffle with pure python code
Best,
Jose

**Comment by takuseno** [comment]
Thank you for your continuous contribution! I'll take a look at this later since I'm busy a bit now.

**Comment by jamartinh** [comment]
Hi @takuseno could you please review this.
I just want to go to another several things and until this is not merged I prefer to wait.
Best,
@jamartinh

File: d3rlpy/base.py [hunk 1]
```diff
@@ -336,15 +340,27 @@ def fit(self,
             if new_transitions:
                 transitions = env_transitions + new_transitions
 
-            indices = np.random.permutation(np.arange(len(transitions)))
```
Comment (takuseno) [inline d3rlpy/base.py:339]: I'd like you to rewrite shuffling like this
```
if shuffle:
    indices = np.random.permutation(np.arange(len(transitions)))
else:
    indices = np.random.arange(len(transitions)
```
Then, the rest of the iterating thins remains the same.

File: d3rlpy/base.py [hunk 2]
```diff
@@ -336,15 +340,27 @@ def fit(self,
...
+                       for i in range(total_batches))
+
+            # Epoch progress bar
+            tqdm_epoch = tqdm(batches,
+                              total=total_batches,
+                              disable=not show_progress,
+                              desc=f'Epoch {epoch + 1}')
+
+            # dict to add incremental mean losses to epoch
+            step_incremental_losses = {}
```
Comment (takuseno) [inline d3rlpy/base.py:361]: In my opinion, this name looks long, which breaks the following line. I would suggest simply `loss_history`.

File: d3rlpy/base.py [hunk 3]
```diff
@@ -336,15 +340,27 @@ def fit(self,
...
+
+            # batches generator
+            total_batches = ((len(transitions) + self.batch_size - 1) //
+                             self.batch_size)
+            batches = (transitions[i * self.batch_size:(i + 1) *
+                                   self.batch_size]
+                       for i in range(total_batches))
+
+            # Epoch progress bar
+            tqdm_epoch = tqdm(batches,
```
Comment (takuseno) [inline d3rlpy/base.py:355]: Since the above comment, I'd like you to use this to replace `range_gen` above.

File: d3rlpy/base.py [hunk 4]
```diff
@@ -288,6 +290,8 @@ def fit(self,
             save_interval (int): interval to save parameters.
             scorers (list(callable)):
                 list of scorer functions used with `eval_episodes`.
+            shuffle (bool): whether or not to shuffle transitions on each
```
Comment (takuseno) [inline d3rlpy/base.py:293]: I'd like to prefer `flag to shuffle transitions on each epoch.`

File: d3rlpy/base.py (hunk 2 above)
Comment (takuseno) [reply d3rlpy/base.py:361]: Or you can add a method to `D3RLPyLogger`, which is like `get_metrics` that returns the current buffer. Then, you can remove `step_incremental_losses`.

**Review by takuseno — CHANGES_REQUESTED** [review]
@jamartinh Thank you for your contribution! And, sorry for the inconvenience and replying late. I've left some comments I'd like you to do. It'd be great if you check those.

File: d3rlpy/base.py (hunk 1 above)
Comment (jamartinh) [reply d3rlpy/base.py:339]: Hi thanks, I think I do not understand, do you mean I should rollback all the batch creation process to keep the building of the list of transitions explicitly without using a generator?

That is to preserve this code?:
```
            n_iters = len(transitions) // self.batch_size
            range_gen = trange(n_iters) if show_progress else range(n_iters)
            for itr in range_gen:
                # pick transitions
                sampled_transitions = []
                head_index = itr * self.batch_size
                for index in indices[head_index:head_index + self.batch_size]:
                    sampled_transitions.append(transitions[index])
```

File: d3rlpy/base.py (hunk 2 above)
Comment (jamartinh) [reply d3rlpy/base.py:361]: ok

File: d3rlpy/base.py (hunk 3 above)
Comment (jamartinh) [reply d3rlpy/base.py:355]: ok

**Review by jamartinh — COMMENTED** [review]
All changes have been done as suggested

**Comment by takuseno** [comment]
@jamartinh LGTM. Thank you for your continuous contribution!!!! I'm really appreciating that. This will be included in the next release. Thanks!

**Review by takuseno — APPROVED** [review]

**Pull request merged**